### ISSUE TITLE
Remove SRP hash from temp storage in redis on success login.

### ISSUE THREAD

**Issue by grempe** [SpiderOak/crypton#407]
See comments in : https://github.com/SpiderOak/crypton/pull/406

**Comment by daviddahl** [comment]
https://github.com/SpiderOak/crypton/blob/master/server/routes/account.js#L145

**Comment by daviddahl** [comment]
Looks like we need to use redisSession.rmProp: https://github.com/SpiderOak/crypton/blob/master/server/redis-session.js#L223

**Comment by grempe** [comment]
So what needs to remain in the redis session key after that?  Here is an example of what is in there now:

````
127.0.0.1:6379> GET Er0mFoM88DRLU94ZhyMFKYICHYCcOSrAnzvTOxX5ilA72pdmZvP88cwUO3OyYlFy
"{\"srpParams\":{\"b\":\"59c520ec8ae084fe85e1cd72dea26d38d91932ac9f3f84364f26c55f725bb3db\",\"B\":\"9bba485dc34320f1b9154fc7e0b177bc0fc2bc60fac59a4a89029bfe24c3f3fbe9d44939491e38682e2e33340121cad156343c7ad022ad94ffbea13ea2e49899ca1becda238121c2e442786874cf55bc32f00dcd2b90d388544abca32e8cd4e3c9bcc9f98a7f3fb61b2b8cb4ff84f23a074ec6ec1b2b6294cfbf4476efef72afe524623bafbb99a052178e31e99c51fe57f94d21369b0e6f901c4932fdc83882ac2856ebaca794bb72660f3a839ae27f100d86a1177d237439fcf607b7d56d60179c3b1779e84d9b456be26bcad60a5341ad67f173fdb98cb2b2feed3efc2fa495767c93b6715a365167a2a5fa5d0f7dc9bfd9621728599ce896a2d5dc474d02\",\"A\":\"43cf412792f01839a1fabb1a4a158279426b3c28de7392c0e872fa9fb9f5cdaddc9405644134d2d88537929944d7a7d8148c61eb1edfb903a8407d457fb6a2852225ce4b6c855f55c3bf049183f9f0a0c78d0b445e940dda5c35ae14905e1a03f2170ef9f50106c164839cab241bb4c792ff01863af9765d84ba9fdbe24c5440d0320444a5bf9f621e9a60fc763679a01e5628ca4e4fb9a3eef6d495ce2b23f6d425d2d36b0c08433761b2108fa9b0823d3c644ed161a6b48e22e02dc78efe691789460a9627a44ad7778e737af6a7e6304f7d6657dfe825bd7dd21cd5a4bc2fdb70b358deb465958175cee9bab1be5307d38fbc91937d857452095bd9d29317\"},\"sid\":\"Er0mFoM88DRLU94ZhyMFKYICHYCcOSrAnzvTOxX5ilA72pdmZvP88cwUO3OyYlFy\",\"accountId\":\"7\"}"
````

So would you just leave 'sid' and 'accountId' object keys?

**Comment by daviddahl** [comment]
(and of course making this module a proper node module as a fork or try to get my changes upstreamed? Maybe?)

**Comment by daviddahl** [comment]
Yes, we only need sid (session ID) and accountId

**Comment by grempe** [comment]
i have not been using node really (I generally am a Ruby person).  So what is the benefit of the home grown redis session code you have vs. something like:

https://github.com/expressjs/session

**Comment by daviddahl** [comment]
We removed all usage of Cookies as this decreases the surfaces we have for exploits

**Comment by grempe** [comment]
While on topic, I'd like to throw something out there to contemplate:

http://macaroons.io
https://github.com/go-macaroon/js-macaroon
http://theory.stanford.edu/%7Eataly/Papers/macaroons.pdf

The thought being to send an opaque signed macaroon to the client (where I think you send a sid now) with the sid embedded and signed (along with possibly other values) and verify on session lookup that the sid you received is signed and valid (within expiry time).  This would be a large barrier against spoofing your sid to the server to gain access to another account.

I say this without having yet done a deep dive on your auth protocol though.  This is along the same line as Javascript Web Tokens (JWT) which are getting a lot of traction but from what I read in the the crypto community macaroons (originally from google I believe) are more respected.

A thought.

**Comment by daviddahl** [comment]
Cool, I will read up on Macaroons.

Yes, JWTs use a flawed protocol. Now, as far as my usage of Crypton is concerned, the session ID and no cookies is secure as I do not use any remote script - we only create signed mobile & desktop apps, not web apps. If you usage is inside web pages, this is probably a wise move due to the remote content and larger potential for abuse.

**Comment by daviddahl** [comment]
Sigh: "This Javascript library (actually, it's written in Typescript)"

**Comment by grempe** [comment]
Well, in both cases your server is still exposed to the network.  And the macaroon verification would always happen on server side.  I don't see the difference whether your client is web or native app.  In both cases you can't trust the client and macaroons would give an additional level of assurance that the incoming request bears an auth token that you know is unforgeable.

Why the issue with the src being Typescript?  The 'compiled' version of the lib that we would use is just JS. Right?

**Comment by daviddahl** [comment]
A signed, packaged app is quite bit more secure than any web page on any server. Full Stop. I am in favor of using somerhing like macaroons to add another layer of defense. My recent research into JWTs was depressing:( 

I am more opposed to CoffeeScript. Yikes! debugging nightmare. I hate introducing things that are hard to debug. A superset of JS like typescript sounds like a nightmare to deal with when things don't work. I like simple to a fault perhaps.

**Comment by grempe** [comment]
Glad you are open to it.

I agree that the use of a signed packaged app is more secure. For the client user only.

Signed packages change nothing about the server side security though. Crypton server just has requests and responses and really knows nothing about the client that is calling it. Crypton doesn't know if its your nice signed packaged app calling it, or a bot calling it and fuzzing every request.  That's why I suggest the macaroon authorization approach.

Totally agree with you that for client side security, and knowing that the code you are executing is legit, you can't beat a signed app with all of its code embedded.  Browsers don't yet allow the same assurances but there is progress being made. For many use cases it can be 'good enough' with the right layers in place (TLS + CSP + HSTS  and now sub-resource integrity)

http://githubengineering.com/subresource-integrity/

But we digress. :-)  Exciting times.

**Comment by daviddahl** [comment]
Your data will be protected because even with a sessionId theft, without your PBKDF2 key an attacker cannot read anything, nor could they even figure out what data to delete as the hmac key makes random item IDs.

However, lets flesh this out a bit. How would macaroons keep bots from creating accounts and calling the server routes? I will need to very soon implement rate limiting at the very least, but I keep thinking about how to prevent bots from creating accounts in the first place, without some horrible "type this unreadabe string into this form" mechanism.

**Comment by grempe** [comment]
Here are some totally off-the-cuff thoughts on how you could take steps to do this. Requires more discussion of course. And a lot more thinking...

The problem of preventing any bogus accounts from being creating (and DOS'ing your box with them) is of course the harder one. I'm wondering if you could think of the account creation itself as being ephemeral unless the user logs in and makes use of the account (which would require also the use of the crypton javascript client).  So essentially, you could allow bots to create an account with a simple POST to /account (with for example a rate limit of one every X amount of seconds), but you would sweep away those accounts that did not follow up within X seconds with creation of a successful session.  For a normal user this would be no problem as they would likely be using your client app to create a login session immediately, since that is what the client does anyway.  Bots however, are likely to not be following the whole protocol and they would never both to login to the account within the specified time period and you could assume they are abandoned accounts and destroy them. This would probably result in deleting nearly 100% of bogus accounts, but only after the fact in a garbage collection step.

Another approach to blocking/slowing account creation would be to require the client to request an account creation 'permission slip' macaroon with an expiring macaroon.  Before they would be allowed access to the POST /account API endpoint you would have to request a permission slip which would only be issued if you had not made more than 1 request to create an account in the last X minutes. Think of it as like a session auth macaroon, but it is limited to use for ONLY the /accounts endpoint. The data in the macaroon is a hash of the IP address (or something?).  You store the hash in redis as well as a counter.  You would then block all requests to /account at the router unless they have this permission macaroon and it is within its expiration time.  Each time they access POST /account you increment the counter in redis. You want to allow a few possible tries but only in the case of requesting a duplicate username. If the account username they requested was new, you allow only a single request per hour (or some time period).

You could also use this technique in combination with the garbage collection one above since this does not 100% prevent creation of bogus accounts. Some smallish number could leak through for any given requesting IP.

Once they login successfully, you invalidate the permission macaroon by deleting its redis key, and you issue a different session macaroon which contains their SID like you do now. All authenticated API endpoints would need a valid session macaroon or they would be immediately rejected with HTTP 401 unauthorized. Every time a session macaroon is presented that is valid, you first check the rate limiter to see if that macaroon SID is within its limits. If so, you fulfill the request. If not you just return 429 Too Many Requests empty response. If the session macaroon looks valid but has expired you tell that to the client with a 401 which then knows it has to go back to the server (without the user even knowing this was needed) and request a new session macaroon.  This would allow macaroons to be relatively short lived (minutes or X # of requests) and prevent their use in a replay attack.

There are probably even better ways to do this.  I may be re-inventing the wheel here.

Side note, the folks at peerio do some interesting things for session/login:
https://github.com/PeerioTechnologies/peerio-client/blob/master/SPECIFICATION.md

**Comment by daviddahl** [comment]
/me reading and thinking, thanks for the copius homework!

**Comment by grempe** [comment]
This issue should be fixed by commit 46d9f00. Please verify correctness.
